### ISSUE TITLE
Out of box set-up fails with an uncaught exception when an existing Cognito IDP is used

### ISSUE THREAD

**Issue by sc0ttdav3y** [aws-observability/aws-rum-web#345]
Hi,

I'm setting up `aws-rum-web` via `npm` into a React Typescript project that has an existing Cognito user pool.

Steps so far:

1. `npm i aws-rum-web`
2. In AWS Console, I followed the bouncing ball to configure a RUM application and pointed it to an existing Cognito user pool + identity pool
3. I added a policy to my "auth" and "unauth" roles in my Cognito IDP with the action `rum:PutRumEvents`
4. I copied the generated JS into my app
5. I've adjusted CSP headers to allow the traffic
6. I've read the READMEs and TROUBLESHOOTING docs

When I run the app, I get this error:

```
Uncaught (in promise) Error: CWR: Failed to retrieve credentials from STS: TypeError: Cannot read properties of undefined (reading 'split')
    at StsClient.eval (StsClient.js?0019:98:1)
    at step (StsClient.js?0019:43:1)
    at Object.eval [as next] (StsClient.js?0019:24:46)
    at fulfilled (StsClient.js?0019:15:43)
```

This is caused by not properly handling an STS error payload, with the root cause to this being at Cognito with something-something basic auth flow. What I see happening is this:

1. `cognito-identity` is called and returns an error such as "Basic flow not enabled" (there are two variants, see below)
2. `sts` is called next with a payload containing `WebIdentityToken: undefined`, so it returns "The ID Token provided is not a valid JWT."
3. The JS in `StsClient.eval()` doesn't correctly parse this response and generates the above error, and suppresses the actual error

The root error to this is about basic auth flow, which I see is part of Cognito Identity Pools.

* If I don't enable Basic Auth Flow, I get the error "Basic (classic) flow is not enabled, please use enhanced flow"
* If I do enable it, I instead get the error "Basic (classic) flow is not supported with RoleMappings, please use enhanced flow"

I've read [the troubleshooting guide](https://github.com/aws-observability/aws-rum-web/blob/main/docs/cdn_troubleshooting.md) which talks to this, and I've also read [open issue 172](https://github.com/aws-observability/aws-rum-web/issues/172) where that user is in a similar position. I've also read of people using proxies, and that'd be great if there's an example I can run with.

However, while [open issue 172](https://github.com/aws-observability/aws-rum-web/issues/172) is asking you to implement the enhanced flow, all I'm after is for the out-of-box experience to work. I figure it works on your machine, so I'm wondering what I've done wrong, or what I need to do in Cognito in order to get this all working?

Many thanks, Scott

**Comment by sc0ttdav3y** [comment]
As an experiment, I set up a new RUM application via the AWS console with defaults, and got it to create a new identity pool as part of its setup. I'd think that would at least get me going, if there's something incompatible with my current Cognito setup.

Steps I followed:

1. In AWS Console > CloudWatch > RUM, click Add App Monitor
2. Set everything as default
3. Copy the revised auth role and application ID into my app
4. Sign out of my app and clear all session data, cookies, etc.

But then I get these network responses in my console, and nothing is successful in RUM.

* "AccessDeniedException:http://internal.amazon.com/coral/com.amazon.rumcontrolplane/" (from dataplane)
* "Invalid identity pool configuration. Check assigned IAM roles for this pool." (from cognito-identity)
* "Caller principal does not have permissions for that resource." (from dataplane)

I'm guessing it's getting confused with my app's Cognito sign-in — which would mean I can't use RUM's identity pool for logging RUM events, along with my app's own identity pool?

To reiterate my initial point, all I'm after is a working out-of-box experience, and I've tried everything I can at this point.

Hope you can point me in the right direction.

Thanks again, Scott

**Comment by qhanam** [comment]
Hi @sc0ttdav3y , if your application's Cognito Identity Pool uses authenticated identities (it probably does), you must create your own `CredentialProvider` and pass it to the RUM web client.

The default Cognito Identity Pool created by RUM contains an "unauthenticated" identity. This means that unauthenticated users can send data to your CloudWatch RUM app monitor. When using the default identity, there is no interaction between your application's authentication and RUM -- it should "just work".

[This article](https://aws.amazon.com/blogs/mt/how-to-isolate-signed-in-users-from-guest-users-within-amazon-cloudwatch-rum/) might help explain CloudWatch RUM's authentication and authorization strategy.

A few things you can try:
- Check that there are no copy/paste errors in the RUM code snippet you copied into your application
- Delete `cwr_c` from localStorage
- Check that the identity pool created by RUM has Cognito's "basic" authentication enabled

**Comment by sc0ttdav3y** [comment]
Thanks for the followup @qhanam. I'll take a read of the linked article.

I got RUM going for now using a second identity pool and anonymous reporting.

But I was with an AWS support engineer and we were unable to get it going with my application's user pool, due to "Basic (classic) flow is not supported with RoleMappings, please use enhanced flow". My application's user and identity pool mirrors the configuration set in Amplify, so I'd expect it to be fairly common. 

I presume all amplify projects will only support anonymous RUM collection at present until the classic/enhanced flow issue is sorted out. Is that your expectation, or would you expect Amplify projects to work with RUM using their own identity pool?

Thanks again... Scott

**Comment by qhanam** [comment]
Incompatibility with enhanced authflow only applies to the unauthenticated identity. Cognito does not currently support RUM for its enhanced authflow when using the unauthenticated identity.

For authenticated identities, you must be pass a credential, or a credential provider, to RUM by calling [`setAwsCredentials`](https://github.com/aws-observability/aws-rum-web/blob/main/docs/cdn_commands.md#commands). Cognito's enhanced authflow can be used in this case.

In [this blog post](https://aws.amazon.com/blogs/mt/how-to-isolate-signed-in-users-from-guest-users-within-amazon-cloudwatch-rum/) we demonstrate an Amplify application using both authenticated and unauthenticated identities with CloudWatch RUM.

**Comment by sc0ttdav3y** [comment]
Thanks @qhanam, that's a useful blog post!

Looks like the TL;DR is to call `setAwsCredentials()` and pass in my user's credentials, and disable the auth role in the RUM config. I'll give that a go and report back.

Thanks again.

**Comment by norchow** [comment]
In case anyone finds it useful, I believe the issue is because we were using multiple app monitors on the same page. 

From the [troubleshooting docs](https://github.com/aws-observability/aws-rum-web/blob/main/docs/cdn_troubleshooting.md#credentials-stored-in-localstorage-are-for-a-different-appmonitor):

> Credentials stored in localStorage are for a different AppMonitor
When anonymous authorization is used, the web client stores credentials in localStorage. If multiple AppMonitors are used within the same domain, authorization will fail when navigating between pages with different AppMonitors.

It seems there's a fix for this implemented in version 1.23.0: https://github.com/aws-observability/aws-rum-web/blob/84b09e204d04b3fc2c9f345cf59d3a096c3d6a4d/src/dispatch/CognitoIdentityClient.ts#L55.

Adding `{cookieAttributes: {unique: true}}` to our config solved the issue.
